dev_tools/nyaml2nxdl/nyaml2nxdl.py

```diff
@@ -0,0 +1,253 @@
+#!/usr/bin/env python3
+"""Main file of nyaml2nxdl tool.
+Users create NeXus instances by writing a YAML file
+which details a hierarchy of data/metadata elements
+
+"""
+# -*- coding: utf-8 -*-
+#
+# Copyright The NOMAD Authors.
+#
+# This file is part of NOMAD. See https://nomad-lab.eu for further info.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import os
+import xml.etree.ElementTree as ET
+
+import click
+
+from .nyaml2nxdl_backward_tools import Nxdl2yaml
+from .nyaml2nxdl_backward_tools import compare_niac_and_my
+from .nyaml2nxdl_forward_tools import nyaml2nxdl
+from .nyaml2nxdl_forward_tools import pretty_print_xml
+from .nyaml2nxdl_helper import extend_yamlfile_with_comment
+from .nyaml2nxdl_helper import get_sha256_hash
+from .nyaml2nxdl_helper import separate_hash_yaml_and_nxdl
+
+DEPTH_SIZE = 4 * " "
+
+# NOTE: Some handful links for nyaml2nxdl converter:
+# https://manual.nexusformat.org/nxdl_desc.html?highlight=optional
+
+
+def generate_nxdl_or_retrieve_nxdl(yaml_file, out_xml_file, verbose):
+    """
+    Generate yaml, nxdl and hash.
+    if the extracted hash is exactly the same as producd from generated yaml then
+    retrieve the nxdl part from provided yaml.
+    Else, generate nxdl from separated yaml with the help of nyaml2nxdl function
+    """
+    pa_path, rel_file = os.path.split(yaml_file)
+    sep_yaml = os.path.join(pa_path, f"temp_{rel_file}")
+    hash_found = separate_hash_yaml_and_nxdl(yaml_file, sep_yaml, out_xml_file)
+
+    if hash_found:
+        gen_hash = get_sha256_hash(sep_yaml)
+        if hash_found == gen_hash:
+            os.remove(sep_yaml)
+            return
+
+    nyaml2nxdl(sep_yaml, out_xml_file, verbose)
+    os.remove(sep_yaml)
+
+
+# pylint: disable=too-many-locals
+def append_yml(input_file, append, verbose):
+    """Append to an existing NeXus base class new elements provided in YML input file \
```

Comment:
As above, short, descriptive, one-line sentence here.  Additional docs as needed but always a blank line after the first when additional docs are presented.

Reply:
:+1:

Reply:
Code gone